op-program/host/host.go

```diff
@@ -142,8 +164,14 @@ func PreimageServer(ctx context.Context, logger log.Logger, cfg *config.Config,
 	hinterDone := routeHints(logger, hintChannel, hinter)
 	select {
 	case err := <-serverDone:
+		// Close the channels to trigger the hinter to exit and wait for it to complete
+		closeChannels()
+		<-hinterDone
```

Comment:
rather than close and wait here. Could we use a defer statement that waits on the hinter/server channels? It's well-defined to receive multiple times from a closed channel.

Reply:
The problem I hit was that `defer` winds up running in the opposite order to what they're called.  So we can't just add another `defer` here because it would wind up executing before the file handles are closed and so we'd deadlock.  But  if we created the `hinterDone` channel here and passed it into `routeHints` so we could setup that `defer` before the file closing defers, it would block forever if we hadn't reached the point where we actually called `routeHints` because  it wouldn't yet be closed.

I kind of hope I'm missing something about how to make this work elegantly though because it's really annoying to not just be able to use `defer` here.

Reply:
One option would be to do:

```
func PreimageServer(ctx context.Context, logger log.Logger, cfg *config.Config, preimageChannel oppio.FileChannel, hintChannel oppio.FileChannel) error {
	var serverDone chan error
	var hinterDone chan error
	defer func() {
		preimageChannel.Close()
		hintChannel.Close()
		if serverDone != nil {
			// Wait for pre-image server to complete
			<-serverDone
		}
		if hinterDone != nil {
			// Wait for hinter to complete
			<-hinterDone
		}
	}()
	logger.Info("Starting preimage server")
```

and then we wouldn't need to manually close and wait like this.  Same pattern works in `FaultProofProgram` as well and I think it's better.  I've pushed up an additional commit with the new approach - let me know what you think.

Reply:
It's still complex/
I tried to come up with a solution that hides complexity in `launchOracleServer` and `routeHints` functions but this cross-synchronization makes it even more complicated.
In the future, I would like to redo I/O to use polling instead of blocking so we can easily handle cancellations.

Reply:
Resolving. This is fine.

Reply:
Yeah agreed, I'm not fan particularly.  Not sure how polling will interact with the VM but worth a look I suspect.